client/consensus/grandpa/environment.go

```diff
@@ -0,0 +1,466 @@
+	// The round number
+	Number uint64
+	// The round state (prevote ghost, estimate, finalized, etc.)
+	State grandpa.RoundState[H, N]
+	// The target block base used for voting in the round
+	Base grandpa.HashNumber[H, N]
+	// All the votes observed in the round
+	// I think this is signature type, double check
+	Votes []grandpa.SignedMessage[H, N, ed25519.SignatureBytes, ed25519.PublicKey]
+}
+
+// NumLastCompletedRounds NOTE: the current strategy for persisting completed rounds is very naive
+// (update everything) and we also rely on cloning to do atomic updates,
+// therefore this value should be kept small for now.
+const NumLastCompletedRounds = 2
+
+// CompletedRounds Data about last completed rounds within a single voter set. Stores
+// NumLastCompletedRounds and always contains data about at least one round
+// (genesis).
+type CompletedRounds[H comparable, N constraints.Unsigned] struct {
+	Rounds []CompletedRound[H, N]
+	SetId  uint64
+	Voters []ed25519.PublicKey
+}
+
+// NewCompletedRounds Create a new completed rounds tracker with NUM_LAST_COMPLETED_ROUNDS capacity.
+func NewCompletedRounds[H comparable, N constraints.Unsigned](genesis *CompletedRound[H, N], setId uint64, voters AuthoritySet[H, N]) CompletedRounds[H, N] {
+	rounds := make([]CompletedRound[H, N], 0, NumLastCompletedRounds)
+	if genesis != nil {
+		rounds = append(rounds, *genesis)
+	}
+
+	var voterIds []ed25519.PublicKey
+	currentAuthorities := voters.CurrentAuthorities
+	for _, auth := range currentAuthorities {
+		voterIds = append(voterIds, auth.Key)
+	}
+
+	return CompletedRounds[H, N]{
+		rounds,
+		setId,
+		voterIds,
+	}
+}
+
+// Last Returns the last (latest) completed round
+func (compRounds *CompletedRounds[H, N]) Last() CompletedRound[H, N] {
+	if len(compRounds.Rounds) == 0 {
+		panic("inner is never empty; always contains at least genesis; qed")
+	}
+	return compRounds.Rounds[0]
+}
+
+// Push a new completed round, oldest round is evicted if number of rounds
+// is higher than `NUM_LAST_COMPLETED_ROUNDS`.
+func (compRounds *CompletedRounds[H, N]) Push(completedRound CompletedRound[H, N]) {
+	// TODO they use reverse, double check this
+	idx, found := slices.BinarySearchFunc(
+		compRounds.Rounds,
+		N(completedRound.Number),
+		func(a CompletedRound[H, N], b N) int {
+			switch {
+			case N(a.Number) == b:
+				return 0
+			case N(a.Number) > b:
+				return 1
+			case N(a.Number) < b:
+				return -1
+			default:
+				panic("huh?")
+			}
+		},
+	)
+
+	if found {
+		compRounds.Rounds[idx] = completedRound
+	} else {
+		if len(compRounds.Rounds) <= idx {
+			compRounds.Rounds = append(compRounds.Rounds, completedRound)
+		} else {
+			compRounds.Rounds = append(compRounds.Rounds[:idx+1], compRounds.Rounds[idx:]...)
+			compRounds.Rounds[idx] = completedRound
+		}
+	}
+
+	if len(compRounds.Rounds) > NumLastCompletedRounds {
+		compRounds.Rounds = compRounds.Rounds[:len(compRounds.Rounds)-1]
+	}
+}
```

Comment:
set info is just a getter for information that is already public I do believe. Regarding iter, since as far as I can tell it is only used in one location and we don't have Iterator objects, I thought it better just to iterate in reverse in the one location where this is called. But can modify if that is not the ideal approach

Reply:
Since they provide the `iter` method and explicitly reverse it we should probably replicate.  You can't expect users to know when or when to not reverse it.

Reply:
updated